### PR TITLE
Initial implementation of `apply_gufunc(..., axes=..., keepdims=...)` (Issue #3843)

### PR DESCRIPTION
**Pull request by magonser**
- [x] Tests added / passed
- [x] Passes `flake8 dask`

This is my attempt for #3843 to implement the `axes` and `keepdims` arguments. It is still work in progress, but I post it, as there may be few things to be discussed.

During implementation some cases appeared that are not straightforward. For the most part, I think they will hardly be used. Nevertheless, here some of the cases - first the easy ones, than the difficult ones:

**Straight forward:**
1) Consuming core dimensions:
    - Examples:
        - `signature="(i)->()", ..., axes=<int>"`
        - `signature="(i),(k)->()", ..., axes=[<int>, <int>]"`
    - Behavior: the respective dimension disappears

2) Passing through core dimensions
    - Example:
        ```python
        a = np.random.randn(4, 5, 6)
        c = apply_gufunc(somefunc, "(i)->(i)", a, axes=1)
        # c.shape is (4, 5, 6]
        ```
    - Behavior:
        - Where should `i` be positioned back in the output? -> I put it in the position of the first array, i.e. that of `a`

**Complicated:**
4) Passing through core dimension for many input array, that had different positions in input. Example:   
```python
        a = np.random.randn(5, 1, 6)
        b = np.random.randn(3, 5)
        c = apply_gufunc(somefunc, "(i),(i)->(i)", a, b, axes=[0, 2])
        # c.shape is (5, 3, 6)
```
- Behavior:
    - Where should `i` be positioned back in the output? -> I put it in the position of the first array, i.e. that of `a`

5) Anything that involves consumed or kept core dimensions and new core coredimensions
    - Example:
        ```python
        a = np.random.randn(3, 4, 5)
        b = np.random.randn(2, 3, 5)
        c = apply_gufunc(somefunc, "(i),()->(k)", a, axes=[(1,), tuple()], keepdims=True, output_sizes={'k': 6})
        # c.shape is (2, 3, 1, 5, 6)
        ```
    - Explanation:
        - dims 0 and 2 are loop dims of `a`, as dim 1 will be put to the end.
        - Hence loop dim shape `(3, 5)` needs to broadcast against `b`which doesn't have any core dims
        - The output has a new core dim, which is always prepended (otherwise we would need a new argument `output_axes`), however the consumed core dimension `i` should be kept. Therefore it is put back in its original position and the new core dim appended.

Yet missing:
- [x] Implement new arguments for `as_gufunc` and `gufunc`
- [ ] Add documentation in `docs/source/array-gufunc.rst`
- [x] Add changelog

### CODE REVIEW

**Comment by magonser** [comment]
@shoyer : any thoughts on the described behaviors? I would like to get some feedback before wrapping it up (documentation only)

**Comment by shoyer** [comment]
Take a look at the NumPy ufunc docs on `axis`/`axes` for details on how the arguments are parsed: https://docs.scipy.org/doc/numpy/reference/ufuncs.html#optional-keyword-arguments

To address your specific examples:

> `c = apply_gufunc(somefunc, "(i)->(i)", a, axes=1)`

This would raise an error because `axes` always needs to be a list. But `axis=1` would work as you describe.

> `c = apply_gufunc(somefunc, "(i),(i)->(i)", a, b, axes=[0, 2])`

This would also raise an error:
`ValueError: axes should be a list with an entry for all 3 inputs and outputs; entries for outputs can only be omitted if none of them has core axes.`

Numba is an easy way to make gufuncs for testing NumPy's behavior, e.g., see https://gist.github.com/shoyer/cce8e042213313685000374200ae498e

**Comment by mrocklin** [comment]
Just checking in here.  How are things going @magonser ?  It looks like you've done some work after @shoyer 's previous comment.  Is this ready for more review or are you still working on things?  If this is ready for more review than @shoyer do you have time sometime next week to give it another look?  (I really appreciate being able to step aside on this particular topic by the way, thank you)

**Comment by magonser** [comment]
I've been meaning to add some more tests in order to be sure the implementation works as intended. 

I am going to state to you explicitly, when ready for review. Thanks.

**Comment by magonser** [comment]
Hi @shoyer 

> This would also raise an error:
> `ValueError: axes should be a list with an entry for all 3 inputs and outputs; entries for outputs can only be omitted if none of them has core axes.`

Do you expect the error to be exactly the same? 

> Numba is an easy way to make gufuncs for testing NumPy's behavior, e.g., see https://gist.github.com/shoyer/cce8e042213313685000374200ae498e

Since Numba is not a dependency to Dask, I have created my own test functions. Any other suggestions?

If both answers to above questions are "No", this PR would be ready for another look.

**Comment by mrocklin** [comment]
> Since Numba is not a dependency to Dask, I have created my own test functions. Any other suggestions?

It would be fine to include numba in tests if you import numba within the test using 

    numba = pytest.importorskip('numba')

**Comment by shoyer** [comment]
> Do you expect the error to be exactly the same?

It's fine to have a different error message as long as we error in the same cases.

I agree with @mrocklin that it's fine to use numba in tests -- we can make it an optional dependency for running tests and include it in most of our Travis-CI builds.

File: dask/array/gufunc.py [hunk 1]
```diff
@@ -171,26 +202,108 @@ def apply_gufunc(func, signature, *args, **kwargs):
     if output_sizes is None:
         output_sizes = {}
 
+    ## Axes
+    ### Input validation
```
Comment (shoyer) [inline dask/array/gufunc.py]: Would it be possible to split some of this logic into another sub-function(s)? `apply_gufunc` is starting to get pretty large.

File: dask/array/gufunc.py [hunk 2]
```diff
@@ -147,7 +174,11 @@ def apply_gufunc(func, signature, *args, **kwargs):
 
     ## Determine and handle output_dtypes
     if output_dtypes is None:
-        output_dtypes = apply_infer_dtype(func, args, kwargs, "apply_gufunc", "output_dtypes", nout)
+        if vectorize:
+            _func = np.vectorize(func, signature=signature)
```
Comment (shoyer) [inline dask/array/gufunc.py]: style nit: I would generally only preface variables with an underscore if it serves to mark a variable as private. In this case, these are already local variables so I would stick to non-prefaced names.

File: dask/array/gufunc.py [hunk 3]
```diff
@@ -171,26 +202,108 @@ def apply_gufunc(func, signature, *args, **kwargs):
     if output_sizes is None:
         output_sizes = {}
 
+    ## Axes
+    ### Input validation
+    if (axes is not None) and (axis is not None):
```
Comment (shoyer) [inline dask/array/gufunc.py]: nit: you don't need parentheses here, `is not` binds more tightly than `and`

File: dask/array/gufunc.py [hunk 4]
```diff
@@ -171,26 +202,108 @@ def apply_gufunc(func, signature, *args, **kwargs):
...
+        raise ValueError("Only one of `axis` or `axes` keyword arguments should be given")
+    if axes and not isinstance(axes, list):
+        raise ValueError("`axes` has to be of type list")
+
+    _core_output_dimss = core_output_dimss if nout else [core_output_dimss]
+    _filtered_core_dims = list(filter(len, core_input_dimss))
+    _nr_outputs_with_coredims = len([True for x in _core_output_dimss if len(x) > 0])
+
+    if keepdims is True:
```
Comment (shoyer) [inline dask/array/gufunc.py]: `keepdims` is already a boolean, so just `if keepdims` is clearer here

File: dask/array/gufunc.py [hunk 5]
```diff
@@ -171,26 +202,108 @@ def apply_gufunc(func, signature, *args, **kwargs):
...
+            axes = [tuple(range(-len(cid), 0)) for cid in _core_dims]
+    elif not isinstance(axes, list):
+        raise ValueError("`axes` argument has to be a list")
+    axes = [(a,) if isinstance(a, int) else a for a in axes]
+
+    ### Treat outputs
+    _nin = len(args)
+    _nout = nout if nout else 1
+
+    if ((_nr_outputs_with_coredims == 0) and (_nin != len(axes)) and (_nin + _nout != len(axes))) \
```
Comment (shoyer) [inline dask/array/gufunc.py]: nit: per PEP8, prefer adding additional parentheses to explicit continuations with `\`

File: dask/array/tests/test_gufunc.py [hunk 6]
```diff
@@ -336,3 +336,237 @@ def foo(x, y):
...
+
+    m = np.median(a, axis=-1 if not axis else axis, keepdims=True)
+    dm = apply_gufunc(mymedian, "(i)->()", da_, axis=axis, keepdims=True, allow_rechunk=True)
+    assert_eq(m, dm)
+
+
+@pytest.mark.parametrize('axes', [[0, 1], [(0,), (1,)]])
+def test_apply_gufunc_axes_01(axes):
+    def mystats(x, y):
+        return np. std(x, axis=-1) * np.mean(y, axis=-1)
```
Comment (shoyer) [inline dask/array/tests/test_gufunc.py]: nit: remove extra space before `std`

File: dask/array/gufunc.py (hunk 1 above)
Comment (magonser) [reply dask/array/gufunc.py]: OK - I've refactored the obvious out in a new function called `_validate_normalize_axes` including the relevant tests.

**Comment by magonser** [comment]
Ready from my side. I would skip on further documentation in `docs/source/array-gufunc.rst` as that is anyway a pretty minimal page so far anyway.

**Comment by mrocklin** [comment]
Merging this in 24 hours if there are no further comments

File: dask/array/gufunc.py [hunk 7]
```diff
@@ -294,14 +321,14 @@ def apply_gufunc(func, signature, *args, **kwargs):
     ## Assess input args for loop dims
     input_shapes = [a.shape for a in args]
     input_chunkss = [a.chunks for a in args]
-    num_loopdims = [len(s) - len(cd) for s, cd in zip(input_shapes, core_input_dimss)]
+    num_loopdims = [len(s) - len(cd) for s, cd in zip(input_shapes, input_coredimss)]
     max_loopdims = max(num_loopdims) if num_loopdims else None
-    _core_input_shapes = [dict(zip(cid, s[n:])) for s, n, cid in zip(input_shapes, num_loopdims, core_input_dimss)]
+    _core_input_shapes = [dict(zip(icd, s[n:])) for s, n, icd in zip(input_shapes, num_loopdims, input_coredimss)]
```
Comment (shoyer) [inline dask/array/gufunc.py]: nit: this should just be `core_input_shapes`

File: dask/array/gufunc.py [hunk 8]
```diff
@@ -55,6 +55,100 @@ def _parse_gufunc_signature(signature):
...
+    nout = 1 if not isinstance(output_coredimss, list) else len(output_coredimss)
+
+    if axes is not None and axis is not None:
+        raise ValueError("Only one of `axis` or `axes` keyword arguments should be given")
+    if axes and not isinstance(axes, list):
+        raise ValueError("`axes` has to be of type list")
+
+    output_coredimss = output_coredimss if nout > 1 else [output_coredimss]
+    _filtered_core_dims = list(filter(len, input_coredimss))
+    _nr_outputs_with_coredims = len([True for x in output_coredimss if len(x) > 0])
```
Comment (shoyer) [inline dask/array/gufunc.py]: nit: these should omit the leading underscores

File: dask/array/gufunc.py [hunk 9]
```diff
@@ -55,6 +55,100 @@ def _parse_gufunc_signature(signature):
     return ins, outs
 
 
+def _validate_normalize_axes(axes, axis, keepdims, input_coredimss, output_coredimss):
```
Comment (shoyer) [inline dask/array/gufunc.py]: thanks for factoring out this logic -- it's much clearer now!

**Comment by mrocklin** [comment]
Merging this on passed tests.  Thanks @magonser for the code and @shoyer for the review!

File: docs/source/changelog.rst [hunk 10]
```diff
@@ -7,7 +7,15 @@ Changelog
 Array
 +++++
 
-- ``apply_gufunc`` implements keywords `axes`, `axis` and `keepdims` (:pr:`3985`) `Markus Gonser`_
```
Comment (magonser) [inline docs/source/changelog.rst]: Just noticed: should we re-include this change?

**Comment by magonser** [comment]
Thanks for the wrap-up work.

On a minor note: merge commit 2354e25 removed the changelog-entry. Should it be re-introduced?

File: docs/source/changelog.rst (hunk 10 above)
Comment (mrocklin) [reply docs/source/changelog.rst]: No.  I now handle this during the release.  I actually slightly prefer that people don't introduce changes themselves.  This tends to cause merge conflicts.

Comment (mrocklin) [reply docs/source/changelog.rst]: Thank you for watching out for this though.

**Pull request merged**